src/distribution_metadata.c

```diff
@@ -411,10 +452,16 @@ IsDistributedTable(Oid tableId)
 
 	SPI_connect();
 
-	spiStatus = SPI_execute_with_args("SELECT NULL "
-									  "FROM pgs_distribution_metadata.partition "
-									  "WHERE relation_id = $1", argCount, argTypes,
-									  argValues, NULL, false, 1);
+	if (spiPlan == NULL)
+	{
+		spiPlan = SPI_prepare("SELECT NULL FROM pgs_distribution_metadata.partition "
+							  "WHERE relation_id = $1", argCount, argTypes);
+
+		spiStatus = SPI_keepplan(spiPlan);
+		Assert(spiStatus == 0);
+	}
+
+	spiStatus = SPI_execute_plan(spiPlan, argValues, NULL, false, 1);
```

Comment:
Just looked into it, and yup:

> When `read_only` is `false`, `SPI_execute` increments the command counter and computes a new _snapshot_ before executing each command in the string. The snapshot does not actually change if the current transaction isolation level is `SERIALIZABLE` or `REPEATABLE READ`, but in `READ COMMITTED` mode the snapshot update allows each command to see the results of newly committed transactions from other sessions. This is essential for consistent behavior when the commands are modifying the database.

> When `read_only` is `true`, `SPI_execute` does not update either the snapshot or the command counter, and it allows only plain `SELECT` commands to appear in the command string. The commands are executed using the snapshot previously established for the surrounding query. This execution mode is somewhat faster than the read/write mode due to eliminating per-command overhead. It also allows genuinely _stable_ functions to be built: since successive executions will all use the same snapshot, there will be no change in the results.

> It is generally unwise to mix read-only and read-write commands within a single function using SPI; that could result in very confusing behavior, since the read-only queries would not see the results of any database updates done by the read-write queries.

I wanted each command to see all committed metadata changes (the locks we have handle making sure consistent boundaries are preserved at a level more fine-grained and semantic than just a whole database snapshot), so the behavior of `read_only` is actually undesirable for us. We want these to act like a bunch of `READ COMMITTED` statements executed one-after-another.